libs/constrained-generators/testlib/Test/Min/Model.hs

```diff
@@ -0,0 +1,1306 @@
+          ]
+      )
+      $
+      -- TODO: we could consider giving `cant` as an argument to `genFromTypeSpec` if this
+      -- starts giving us trouble.
+      genFromTypeSpec s `suchThatT` (`notElem` cant)
+  ErrorSpec e -> genErrorNE e
+
+-- | A version of `genFromSpecT` that simply errors if the generator fails
+genFromSpec :: forall a. (HasCallStack, HasSpec a) => Spec a -> Gen a
+genFromSpec spec = do
+  res <- catchGen $ genFromSpecT @a @GE spec
+  either (error . ('\n' :) . catMessages) pure res
+
+-- | A version of `genFromSpecT` that runs in the IO monad. Good for debugging.
+debugSpec :: forall a. HasSpec a => Spec a -> IO ()
+debugSpec spec = do
+  ans <- generate $ genFromGenT $ inspect (genFromSpecT spec)
+  let f x = putStrLn (unlines (NE.toList x))
+      ok x =
+        if conformsToSpec x spec
+          then putStrLn "True"
+          else putStrLn "False, perhaps there is an unsafeExists in the spec?"
+  case ans of
+    FatalError xs -> mapM_ f xs
+    GenError xs -> mapM_ f xs
+    Result x -> print spec >> print (simplifySpec spec) >> print x >> ok x
+
+-- | Generate a satisfying `Env` for a `p : Pred fn`. The `Env` contains values for
+-- all the free variables in `flattenPred p`.
+genFromPreds :: forall m. MonadGenError m => Env -> Pred -> GenT m Env
+-- TODO: remove this once optimisePred does a proper fixpoint computation
+genFromPreds env0 (optimisePred . optimisePred -> preds) =
+  {- explain1 (show $ "genFromPreds fails\nPreds are:" /> pretty preds) -} do
+    -- NOTE: this is just lazy enough that the work of flattening,
+    -- computing dependencies, and linearizing is memoized in
+    -- properties that use `genFromPreds`.
+    plan <- runGE $ prepareLinearization preds
+    go env0 plan
+  where
+    go :: Env -> SolverPlan -> GenT m Env
+    go env plan | isEmptyPlan plan = pure env
+    go env plan = explain (show $ "Stepping the plan:" /> vsep [pretty env, pretty (substPlan env plan)]) $ do
+      (env', plan') <- stepPlan env plan
+      go env' plan'
+
+-- =============================================================
+-- Simplifcation
+-- =============================================================
+
+simplifySpec :: HasSpec a => Spec a -> Spec a
+simplifySpec spec = case regularizeNames spec of
+  SuspendedSpec x p ->
+    let optP = optimisePred p
+     in fromGESpec $
+          explain
+            ("\nWhile calling simplifySpec on var " ++ show x ++ "\noptP=\n" ++ show optP ++ "\n")
+            (computeSpecSimplified x optP)
+  MemberSpec xs -> MemberSpec xs
+  ErrorSpec es -> ErrorSpec es
+  TypeSpec ts cant -> TypeSpec ts cant
+  TrueSpec -> TrueSpec
+
+-- | Turn 'GenError' into 'ErrorSpec', and FatalError into 'error'
+fromGESpec :: HasCallStack => GE (Spec a) -> Spec a
+fromGESpec ge = case ge of
+  Result s -> s
+  GenError xs -> ErrorSpec (catMessageList xs)
+  FatalError es -> error $ catMessages es
+
+------- Stages of simplifying -------------------------------
```

Comment:
It's a shame we need a bunch of this stuff for this "simplified" version of the system to work. I'd like to sit down once we've shrunk the code-base here some more and see if I can get rid of some of this.

Reply:
The problem is that making a SolverPlan, requires all this stuff. We don't talk about how we do that, only what the result looks like, and how it is used.